internal/pkg/encoding/jsonnet/validate.go

```diff
@@ -0,0 +1,17 @@
+package jsonnet
+
+import (
+	"github.com/google/go-jsonnet/ast"
+	"github.com/google/go-jsonnet/parser"
+)
+
+func (c *Context) Validate(code string) error {
+	// get global vars
+	vars := make([]ast.Identifier, 0, len(c.globalBinding))
+	for k := range c.globalBinding {
+		vars = append(vars, k)
+	}
+
+	_, err := parser.SnippetToAst(code, c.FilePath(), vars...)
```

Comment:
`SnippetToAst` pošle code skrz zákládní analýzu, mezi kterou patří i detekování undefined variables. Trvalo mi to dýl, protože jsem si myslel, že na tohle není public API a začal jsem to psát manuálně podle vzoru toho, co dělají oni (managing scopes + declare/resolve variables). Až potom jsem narazil na to, že to je vlastně součástí public API 🙂. V dokumentaci o tom není ani zmínka.

Reply:
Imho by nebylo zas tak moc práce napsat nějaký primitivní unification pass s hindley-milner type inference (+row polymorphism), protože ten jazyk je tak minimální (funkce, objekty, pole + pár primitivních typů) a mít tím pádem i static type checking (bez type anotací) a lepší error messages. Něco podobného jsem napsal [zde](https://github.com/ezclap-tv/mu-lang/blob/type-system-experiment/experiments/type-system/tests/infer.rs) a je to nějakých ~300 řádků. Nevím ale proč to neudělali sami, možná je tam nějaký fundamentální blocker 🙂.

Reply:
Na lepsie error msgs maju tusim na GitHub issue. Niekde som take nieco videl.

Reply:
> static type checking 
to my myslim nefungovalo u tych `native go functions`, ktore mozu vratit hocico.

Reply:
Tam stačí je registrovat před spuštením type checkingu a otypovat je, zas takový problém to není. V Go by dokonce šlo automaticky udělat inference těch nativních funkcí pomocí reflect package.

Reply:
No v nasom pripade je tam `any`, kedze request JSON `Body` moze obsahovat hocico.

Reply:
To je v pořádku, ono hlavně jde o to zachytit ty "obvious" chyby, jako třeba zavolání funkce se špatným počtem args. `any` je trochu "díra" v type systému, ale ten nemusí být kompletní na to, aby byl užitečný.